ability to clean WITH clause matching CONTAIN process

mikeweb85 opened the pull request:
Added NULL condition to overwrite and clear SQL WITH clause matching the CONTAIN  process.

Commit:
ability to clean WITH clause matching CONTAIN process

mikeweb85 committed:
Merge branch '5.x' into 5.x

dereuromark commented:
Ideally this should come with a test case.

ndm2 commented on src/Database/Query.php:
```diff
@@ -402,7 +402,10 @@ public function with(CommonTableExpression|Closure $cte, bool $overwrite = false
         }
 
-        $this->_parts['with'][] = $cte;
+        if ($cte) {
+            $this->_parts['with'][] = $cte;
+        }
```
Not sure if a no-op like `with(null)` makes sense, but other query builder methods with override support allow similar no-ops.

ndm2 commented:
Technically this would need to go into 6.x, as the signature change is not backwards compatible. If it's considered low risk, then maybe it could go into 5.next.

markstory commented:
> Technically this would need to go into 6.x, as the signature change is not backwards compatible.

Why isn't it though? All call sites that were valid before are still valid. The addition is new semantic meaning to a newly supported type. We don't have a clear example of 'adding nullability' in our [backwards compatibility guide](https://book.cakephp.org/5/en/contributing/backwards-compatibility.html#classes) but perhaps it is time we decide on that.

I agree on this going to 5.next though as it isn't a bugfix.

markstory committed:
Add test for #17563

Currently the test is failing, but perhaps I don't understand the use
case well enough.

markstory commented:
I've added a test, and I'm not sure I understand the workflow this change would enable. Any examples you could share?

mikeweb85 commented:
sorry, been I'll. yes, I have an example that prompted this tweak. I'll drop that here in the morning.

ndm2 commented:
> > Technically this would need to go into 6.x, as the signature change is not backwards compatible.
> 
> Why isn't it though? All call sites that were valid before are still valid. The addition is new semantic meaning to a newly supported type. We don't have a clear example of 'adding nullability' in our [backwards compatibility guide](https://book.cakephp.org/5/en/contributing/backwards-compatibility.html#classes) but perhaps it is time we decide on that.
> 
> I agree on this going to 5.next though as it isn't a bugfix.

Now that we have actually typed arguments, adding a type (`null` or otherwise) would break all overridden methods, and so far breaking a public API in this was always considered not BC?

mikeweb85 commented:
here is an example for where/why resetting the CTE was useful. Existing paginator has memory issues with large MSSQL datasets.
![image](https://github.com/cakephp/cakephp/assets/3333004/acbeb4fa-f234-4368-8fc2-0171b9aaeaab)

markstory committed:
Update test to overwrite the with clause

dereuromark commented:
@markstory 
Anything blocking this?

markstory commented:
> Anything blocking this?

Not really. However, it is currently targetting `5.x`. Is this too much new behavior for a bugfix release?

othercorey commented:
Should we change this to support an array of cte instead of a nullable cte? Would that align more with other array parts?

ADmad commented:
> Should we change this to support an array of cte instead of a nullable cte? Would that align more with other array parts?

Given that the `WITH` clause can have multiple CTEs and internally also it's also stored as an array, allowing array as argument makes sense.

othercorey commented:
replaced by https://github.com/cakephp/cakephp/pull/18706
